src/core/service.c

```diff
@@ -262,6 +265,27 @@ static void service_start_watchdog(Service *s) {
                 log_unit_warning_errno(UNIT(s), r, "Failed to install watchdog timer: %m");
 }
 
+static usec_t service_set_restart_usec(Service *s) {
+        long double unit;
+
+        assert(s);
+
+        /* This is called before n_restarts is updated */
+        if (s->n_restarts == 0 ||
+            s->restart_steps == 0 ||
+            s->restart_usec_max == USEC_INFINITY ||
+            s->restart_usec == s->restart_usec_max)
+                return s->restart_usec;
+
+        if (s->n_restarts >= s->restart_steps)
+                return s->restart_usec_max;
+
+        unit = powl(s->restart_usec_max - s->restart_usec, 1.0L / s->restart_steps);
```

Comment:
add an assert here that restartusec_max really is >= restart_usec. (and maybe a comment that this is enforced in service_verify()